book/src/types/rust_types.md

```diff
@@ -0,0 +1,246 @@
+answeres for `exists<T> { Implemented(T: Trait) }`, then we know that
+the negation is true. (Note that this means that e.g. `forall<X> { X =
+i32 }` is false but so is `forall<X> { not { X = i32 } }`.)
+
+### Inference variables
+
+The `InferenceVar` variant wraps an `InferenceVar` type.  This
+represents a type whose value is being inferred. The value of an
+inference variables may be "known" or "not known", but that state is
+stored externally, in the inference context (see the section on
+inference below).
+
+When equating, inference variables are treated specially in that they
+become bound (or, if they have already been bound, they are replaced
+with their value).
+
+Inference variables are also integral to canonicalization and
+other types.
+
+### Dyn types
+
+The `Dyn` variant wraps a `DynTy` and represents a `dyn Trait`
+type. In chalk, these are represented as an existential type where we
+store the predicates that are known to be true. So a type like `dyn
+Write` would be represented as, effectively, an `exists<T> { T: Write
+}` type.
+  
+When equating, two `dyn P` and `dyn Q` types are equal if `P = Q` --
+i.e., they have the same bounds. Note that -- for this purpose --
+ordering of bounds is significant, and chalk assumes that a canonical
+ordering is established from somewhere else.
+
+**Question:** We should consider whether logical entailment (e.g., `P
+=> Q` and `Q => P`) would be a better basis. This is more of a Rust
+language design question.
+
+There are "automatic" rules for proving that `dyn P: P` and so forth, but
+that is outside the scope of the chalk-ir crate.
+
+*Question:* in rustc, `dyn Foo` has a lifetime bound, I guess we need that in chalk?
+
+### Function pointer types
+
+The `Fn` variant wraps a `FnTy` struct and represents a `fn()` type
+(in other words, a function pointer). In some ways, fn types are like
+application types, but with one crucial difference: they also contain
+a `forall` binder that for lifetimes whose value is determined when
+the function is called. Consider e.g. a type like `fn(&u32)` or --
+more explicitly -- `for<'a> fn(&'a u32)`.
+  
+Two `Fn` types `A, B` are equal `A = B` if `A <: B` and `B <: A`
+
+Two `Fn` types `A, B` are subtypes `A <: B` if
+
+* After instantiating the lifetime parameters on `B` universally...
+    * You can instantiate the lifetime parameters on `A` existentially...
+        * And then you find that `P_B <: P_A` for every parameter type `P` on `A` and `B` and
+          `R_A <: R_B` for the return type `R` of `A` and `B`.
+          
+We currently handle type inference with a bit of a hack (same as
+rustc); when relating a `Fn` type `F` to an unbounded type
+variable `V`, we instantiate `V` with `F`.  But in practice
+because of the above subtyping rules there are actually a range of
+values that `V` could have and still be equal with `F`. This may
+or may not be someting to consider revisiting.
+
+### Generator witness types
```

Comment:
Unfortunately, documentation for these seems lacking. Any links to any documentation of these in rustc?

Reply:
Not sure, there may be some comments in rustc. I'm not aware of any section in the rustc-guide. There *should* be, I agree. =) Work to do!